### PR TITLE
Reduce audio lags while accessing smm

### PR DESCRIPTION
**Pull request by lvs1974**
Here is my next attempt to reduce audio lags.
Idea: hook method IOAudioStream::processOutputSamples(IOAudioStream *this, void *clientBuffer, UInt32 firstSampleFrame, UInt32 loopCount, bool samplesAvailable) and do not read SMM while there are any processed audio samples.

### CODE REVIEW

**Comment by lvs1974** [comment]
Here is the kext for testing:
[SMCDellSensors.kext.zip](https://github.com/acidanthera/VirtualSMC/files/5323762/SMCDellSensors.kext.zip)


@tctien342, @naveenkrdy: could you test, please?

File: Sensors/SMCDellSensors/SMIMonitor.cpp [hunk 1]
```diff
@@ -9,21 +9,31 @@
...
+#include "kern_hooks.hpp"
+
 #include <Headers/kern_cpu.hpp>
 
 extern "C" {
 #include <i386/pmCPU.h>
 }
 
 SMIMonitor *SMIMonitor::instance = nullptr;
+_Atomic(bool) volatile SMIMonitor::smmIsBeingRead = false;
```
Comment (vit9696) [inline Sensors/SMCDellSensors/SMIMonitor.cpp]: What is volatile atomic? @_@

**Comment by 07151129** [comment]
I haven't looked at the code yet, but I suggest another approach: assign a low priority to the sensor readout thread. This should give some extra time to the audio threads. I know some of them already have real-time mach priority, but not sure which ones exactly.

**Comment by lvs1974** [comment]
@07151129: I have already tried to use the lowest priority for sensor readout thread, an it had no any effect. Lower priority does not prevent audio stream from processing audio samples (smm reading can be already started by this mpment), quite the contrary. But I could try to use both: implementation in this PR and decrease priority.

**Comment by lvs1974** [comment]
@07151129: I have tried to reduce priority and it got worse.

File: Sensors/SMCDellSensors/SMIMonitor.cpp [hunk 2]
```diff
@@ -9,21 +9,31 @@
...
 
 OSDefineMetaClassAndStructors(SMIMonitor, OSObject)
 
 int SMIMonitor::i8k_smm(SMMRegisters *regs) {
 	int rc;
 	int eax = regs->eax;  //input value
 	
-	IOSimpleLockLock(preemptionLock);
+	smmIsBeingRead = true;
```
Comment (vit9696) [inline Sensors/SMCDellSensors/SMIMonitor.cpp]: Use atomic_load/store_explicit here and everywhere else and drop volatile? If you do not want to bother about memory ordering could just use the sequential consistency for now.

**Comment by tctien342** [comment]
@lvs1974 so far so good man, testing 3 hours with BT speaker, not have any delay!

**Comment by lvs1974** [comment]
@vit9696: could you check a usage of memory_order constants for atomic_load_explicit / atomic_store_explicit / atomic_fetch_add_explicit / atomic_fetch_sub_explicit, please?

**Comment by vit9696** [comment]
Last change to remove IOSleep, will not it cause high power usage due to busy loops?

**Comment by lvs1974** [comment]
> Last change to remove IOSleep, will not it cause high power usage due to busy loops?

On my pc it does not cause high power usage.
But, actually, it make no difference, so the last commit can be removed.

**Comment by tctien342** [comment]
Comfirmed latest commit work great with my AptX Headphone!
+ OS: Big sur beta 10
+ Dell Ins7591

**Pull request merged**